### PR TITLE
Fix `x` behavior on empty lines

### PR DESCRIPTION
**Pull request by dpc**
This is to address my personal issue with #356.

I'm not sure if this is an ideal implementation, or will it get merged, but was easy enough and it should be easy for me to use a personal fork, just like I used to for kakoune: https://github.com/dpc/kakoune-dpc/

### CODE REVIEW

**Comment by dpc** [comment]
I can use it in my NixOS with:

```
   (helix.overrideAttrs ( old: rec {
      pname = "helix";
      version = "c732b9c3369798837b05a1b3a4d728ed82abe461";
      src = fetchFromGitHub {
        repo = "helix";
        owner = "dpc";
        rev = "${version}";
        sha256 = "sha256-bJJN6Rt4gIZDEOiujlNrY57ylVsfWXABiNRC0u8qLWs=";
      };

      cargoDeps = old.cargoDeps.overrideAttrs (lib.const {
        name = "${pname}-vendor.tar.gz";
        inherit src;
        outputHash = "sha256-LNN6ix0VObiwDYr+4LaLFSTKU+WqcPguFQt/l3LfgfI=";
      });
    }))
```

File: helix-core/src/selection.rs [hunk 1]
```diff
@@ -54,6 +54,9 @@ pub struct Range {
     /// The head of the range, moved when extending.
     pub head: usize,
     pub horiz: Option<u32>,
+    /// True if range is intended to contain whole lines (e.g. after `line_extend_below`).
+    /// Useful to prevent first `line_extend_below` on an empty line selecting the next line.
+    pub line_mode: bool,
```
Comment (archseer) [inline helix-core/src/selection.rs:59]: This should be a field on the Selection, not Range?

Comment (dpc) [reply helix-core/src/selection.rs:59]: I have no idea. :) . Seem to work OK for me, AFAICT.

Comment (dpc) [reply helix-core/src/selection.rs:59]: I actually just tried moving `line_mode` to selection, and it doesn't work well there. AFAICT, "already selected whole lines" is more of a property of a range withing selection, than whole selection.

**Comment by apoelstra** [comment]
For people finding this later -- this PR is old of course but it looks like dpc is actively maintaining his patch on https://github.com/dpc/helix

You can apply the current one on NixOS by adding to your systempackages

```
    (helix.overrideAttrs (old: {
      patches = (old.patches or []) ++ [
        (fetchpatch {
          name = "dpc-fix-x.patch";
          url = "https://github.com/helix-editor/helix/commit/054c33f530168e9fc61940a2a9e441debf6a73cc.patch";
          hash = "sha256-wvjzIHdi2ZH8lKhqAVxLP1xcjtjr92DGXMU8nZY9CkM=";
        })
      ];
    }))
```

in place of helix.

**Comment by dpc** [comment]
Surprisingly, I don't think this patch required much "maintaining". It keeps applying cleanly for more than 2 years now, IIRC.

**Comment by eugenesvk** [comment]
Could this maybe be merged as a separate command if the default intuition-tripping behavior must be preserved?

**Comment by dpc** [comment]
BTW. I'll probably personally switch to: https://github.com/helix-editor/helix/discussions/15172

**Pull request closed without merging**